File: src/main/java/axirassa/overlord/Overlord.java

package axirassa.overlord;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import axirassa.overlord.NoOverlordConfigurationException;
import axirassa.overlord.OverlordException;
import axirassa.overlord.UnknownExecutionTargetException;
import axirassa.overlord.os.AbstractOverlordSystemSupport;
import axirassa.overlord.os.OverlordSystemSupport;

/**
 * A process starting/monitoring daemon and framework.
 * 
 * AxOverlord hooks into a HornetQ message passing framework.
 * 
 * Note that java processes started by Overlord will not be terminated by
 * default unless {@link #addShutdownHooks()} is called.
 * 
 * @author wiktor
 * 
 */
public class Overlord {
	public static final Logger logger = LoggerFactory.getLogger(Overlord.class);
	private static final String CONFIGURATION_FILE = "axoverlord.cfg.xml";

	private Integer execId = 0;


	public int getNextExecID () {
		synchronized (execId) {
			execId++;
			return execId;
		}
	}


	public static void main (String[] parameters) throws OverlordException, IOException, InterruptedException {
		Overlord overlord = new Overlord();
		overlord.addShutdownHooks();

		if (parameters.length <= 0)
			overlord.execute(new String[] { "master" });
		else
			overlord.execute(parameters);
	}


	//
	// Class Instances
	//
	private OverlordConfiguration configuration;
	private final ArrayList<ExecutionInstance> instances = new ArrayList<ExecutionInstance>();
	private final NativeLibraryProvider libprovider = new NativeLibraryProvider();


	public void execute (String[] parameters) throws OverlordException, IOException, InterruptedException {
		OverlordSystemSupport systemsupport = AbstractOverlordSystemSupport.getSystemSupport();

		URL configfile = ClassLoader.getSystemResource(CONFIGURATION_FILE);
		if (configfile == null)
			throw new NoOverlordConfigurationException(CONFIGURATION_FILE);
		InputStream configstream = ClassLoader.getSystemResourceAsStream(CONFIGURATION_FILE);

		configuration = new OverlordConfiguration(this);
		configuration.setJavaExecutable(systemsupport.getJavaExecutable());

		XMLConfigurationParser configparser = new XMLConfigurationParser(configfile, configstream, configuration);
		configparser.parse();

		List<ExecutionGroup> groups = new ArrayList<ExecutionGroup>();

		for (String groupname : parameters) {
			ExecutionGroup group;
			if (groupname.matches("t:.*")) {
				String targetName = groupname.substring(2);

				ExecutionTarget target = configuration.getExecutionTarget(targetName);
				if (target == null)
					throw new UnknownExecutionTargetException(targetName, null);

				ExecutionSpecification spec = new ExecutionSpecification(configuration, target);
				spec.setInstances(1);

				group = new ExecutionGroup("target_" + targetName);
				group.addExecutionSpecification(spec);
			} else {
				group = configuration.getExecutionGroup(groupname);
			}

			if (group != null)
				groups.add(group);
			else {
				logger.error("Unknown Execution Group: {}", group);
				return;
			}
		}

		for (ExecutionGroup group : groups)
			group.execute();
	}


	public Collection<ExecutionInstance> getExecutionInstances () {
		return instances;
	}


	public NativeLibraryProvider getNativeLibraryProvider () {
		return libprovider;
	}


	public void addShutdownHooks () {
		Runtime.getRuntime().addShutdownHook(new OverlordDynamicShutdownHook(this));
	}


	public void addExecutionInstance (Thread thread, ExecutionMonitor monitor) {
		instances.add(new ExecutionInstance(thread, monitor));
	}


	public void killInstances () {
		for (ExecutionInstance instance : instances)
			if (instance.getThread().isAlive()) {
				instance.getThread().interrupt();
				instance.getMonitor().killProcess();
			}
	}
}

class ExecutionInstance {
	private final Thread thread;
	private final ExecutionMonitor monitor;


	public ExecutionInstance (Thread thread, ExecutionMonitor monitor) {
		this.thread = thread;
		this.monitor = monitor;
	}


	public Thread getThread () {
		return thread;
	}


	public ExecutionMonitor getMonitor () {
		return monitor;
	}
}

File: src/main/java/axirassa/overlord/Overlord.scala

package axirassa.overlord

import java.util.ArrayList
import org.slf4j.LoggerFactory

import scala.collection.JavaConversions._

/**
 * A process starting/monitoring daemon and framework.
 *
 * AxOverlord hooks into a HornetQ message passing framework.
 *
 * Note that java processes started by Overlord will not be terminated by
 * default unless {@link #addShutdownHooks()} is called.
 *
 * @author wiktor
 *
 */
object Overlord {
  val CONFIGURATION_FILE = "axoverlord.cfg.xml"

  def main(arg : Array[String]) {
    val overlord = new Overlord()
    overlord.addShutdownHooks()

    if (arg.length <= 0)
      overlord.execute(Array("master"))
    else
      overlord.execute(arg)
  }

  var execId = 0

  def getNextExecID() = {
    synchronized {
      execId += 1
      execId
    }
  }
}

class Overlord {
  val logger = LoggerFactory.getLogger(classOf[Overlord])

  //
  // Class Instances
  //

  var configuration : OverlordConfiguration = _
  var instances = new ArrayList[ExecutionInstance]
  var libraryProvider = new NativeLibraryProvider

  def execute(parameters : Array[String]) {
    val systemSupport = OverlordSystemSupport.getSystemSupport

    val configFile = ClassLoader.getSystemResource(Overlord.CONFIGURATION_FILE)
    if (configFile == null)
      throw new NoOverlordConfigurationException(Overlord.CONFIGURATION_FILE)
    val configStream = ClassLoader.getSystemResourceAsStream(Overlord.CONFIGURATION_FILE)

    println("CONFIGURATION FILE: "+configFile)
    println("CONFIGURATION STREAM: "+configStream)

    configuration = new OverlordConfiguration(this)
    configuration.javaExecutable = systemSupport.javaExecutable

    val configParser = new XMLConfigurationParser(configFile, configStream, configuration)
    configParser.parse()

    val groups = new ArrayList[ExecutionGroup]()

    for (groupName <- parameters) {
      var group : ExecutionGroup = null
      if (groupName.matches("t:.*")) {
        val targetName = groupName.substring(2)

        val target = configuration.getExecutionTarget(targetName)
        if (target == null)
          throw new UnknownExecutionTargetException(targetName, null)

        val spec = new ExecutionSpecification(configuration, target)
        spec.instances = 1

        group = new ExecutionGroup("target_"+targetName)
        group.addExecutionSpecification(spec)
      } else {
        group = configuration.getExecutionGroup(groupName)
      }

      if (group != null)
        groups.add(group)
      else {
        logger.error("Unknown Execution Group: {}", group)
        return
      }
    }

    for (group <- groups)
      group.execute()
  }

  def addShutdownHooks() {
    Runtime.getRuntime().addShutdownHook(new OverlordDynamicShutdownHook(this))
  }

  def addExecutionInstance(thread : Thread, monitor : ExecutionMonitor) {
    instances.add(new ExecutionInstance(thread, monitor))
  }

  def killInstances() {
    for (instance <- instances)
      if (instance.thread.isAlive()) {
        instance.thread.interrupt()
        instance.monitor.killProcess()
      }
  }
}

class ExecutionInstance(val thread : Thread, val monitor : ExecutionMonitor)